### DISCUSSION TITLE
Unexpected compile error when accessing URL props via super.prop [node]

### DISCUSSION THREAD

**Discussion by LeviticusMB** [DefinitelyTyped/DefinitelyTyped#70598] [Issues with a @types package]
I recently raised an issue in microsoft/TypeScript#59892 where the compiler (since 5.3) incorrectly errors out when using `super` to access props from a subclass of `URL`. This only happens when using the `URL` definitions from `@types/node`, not `lib.dom.d`.

The conclusion was that `@types/node` ought to define the `URL` props using getters and setters and not as normal class instance props.

Please have a look at the linked issue and the [demo repo](https://github.com/LeviticusMB/demo-ts-super).

[a comment by an automated account is not shown]

**Comment by Renegade334** [comment]
To continue the discussion from the linked issue:

Web interface definitions have "attributes" in the web standard, which are actually getters (for readonly attributes) or getters/setters (for writable attributes), but which are emitted as instance properties by TypeScript-DOM-lib-generator. For instance, the URL interface in the TypeScript DOM library looks like:
```ts
interface URL {
    hash: string;
    host: string;
    hostname: string;
    href: string;
    ...
}
declare var URL: {
    prototype: URL;
    new(url: string | URL, base?: string | URL): URL;
    ...
}
```
but inspecting `URL.prototype` in your engine of choice will show that they're not actually instance properties, but getters and setters defined on the prototype:
```js
>> URL.prototype
URLPrototype { toJSON: toJSON(), toString: toString(), href: Getter & Setter, origin: Getter, … }
```

@types/node is consistent with the TypeScript library, in that it also defines these attributes as instance properties. The difference is that in @types/node, as in the Node runtime, these web interfaces are defined as classes rather than newable functions:
```ts
class URL {
    constructor(input: string | { toString: () => string }, base?: string | URL);
    hash: string;
    host: string;
    hostname: string;
    href: string;
    ...
}
```

The difference in TypeScript that prompted the initial issue is that the compiler (as at 5.6) doesn't perform `super.*` accessibility checks on ES5-style constructors, only on classes. That means that something like
```ts
class MyURL extends URL {
    constructor(input: string, base?: string | URL) {
        super(input, base);
        super.host; // <========
    }
}
```
– which, according to the types, shouldn't work in either case – will only raise a compilation error with the @types/node definitions, not with TypeScript's lib.dom definitions.

The question raised is whether @types/node should change its definitions so that web interface attributes are defined as getters/setters rather than instance properties, to allow this specific behaviour (pseudoproperty access via `super`). My personal opinion is that @types/node should continue to mirror the approach used in TypeScript's DOM definitions, and therefore keep the status quo unless this change is made within the TypeScript library: it's a large change to facilitate a very niche behaviour, and judicious use of `@ts-ignore` is probably justifiable here.
